src/vm/collection.rs

```diff
@@ -63,13 +64,22 @@ pub trait Collection<VM: VMBinding> {
         m: &T,
     );
 
-    /// Inform the VM for an out-of-memory error. The VM can implement its own error routine for OOM.
-    /// Note the VM needs to fail in this call. We do not expect the VM to resume in any way.
+    /// Inform the VM of an out-of-memory error. The binding should hook into the VM's error
+    /// routine for OOM. Note that there are two different categories of OOM:
+    ///  * Critical OOM: This is the case where the OS is unable to mmap or acquire more memory.
+    ///    MMTk expects the VM to abort immediately if such an error is thrown.
+    ///  * Heap OOM: This is the case where the specified heap size is insufficient to execute the
+    ///    application. MMTk expects the binding to notify the VM about this OOM and throw an OOM
+    ///    error *at some point later in time*. That is to say, MMTk does not expect the VM to
```

Comment:
We can leave the choice to the VM whether they want to return or panic.

Reply:
Yes I will make it more explicit.